Bug: bazel running out of memory and crashing on a large project

## Describe the bug

When I run the `xcodeproj` rule on an iOS app with a large dependency graph, bazel crashes almost immediately with the following trace:

```
FATAL: bazel ran out of memory and crashed. Printing stack trace:
net.starlark.java.eval.Starlark$UncheckedEvalError: OutOfMemoryError thrown during Starlark evaluation (@envoy//source/common/upstream:cds_api_helper_lib)
	at <starlark>.encode(<builtin>:0)
	at <starlark>._xcode_target(/private/var/tmp/_bazel_jsimard/7140f038a5c13b01a171c93151a000da/external/com_github_buildbuddy_io_rules_xcodeproj/xcodeproj/internal/target.bzl:281)
	at <starlark>._process_library_target(/private/var/tmp/_bazel_jsimard/7140f038a5c13b01a171c93151a000da/external/com_github_buildbuddy_io_rules_xcodeproj/xcodeproj/internal/target.bzl:806)
	at <starlark>._process_target(/private/var/tmp/_bazel_jsimard/7140f038a5c13b01a171c93151a000da/external/com_github_buildbuddy_io_rules_xcodeproj/xcodeproj/internal/target.bzl:1430)
	at <starlark>.process_target(/private/var/tmp/_bazel_jsimard/7140f038a5c13b01a171c93151a000da/external/com_github_buildbuddy_io_rules_xcodeproj/xcodeproj/internal/target.bzl:1495)
	at <starlark>._xcodeproj_aspect_impl(/private/var/tmp/_bazel_jsimard/7140f038a5c13b01a171c93151a000da/external/com_github_buildbuddy_io_rules_xcodeproj/xcodeproj/internal/xcodeproj_aspect.bzl:40)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.Arrays.copyOf(Unknown Source)
	at java.base/java.lang.AbstractStringBuilder.ensureCapacityInternal(Unknown Source)
	at java.base/java.lang.AbstractStringBuilder.append(Unknown Source)
	at java.base/java.lang.StringBuilder.append(Unknown Source)
	at net.starlark.java.lib.json.Json$Encoder.appendQuoted(Json.java:260)
	at net.starlark.java.lib.json.Json$Encoder.encode(Json.java:114)
	at net.starlark.java.lib.json.Json$Encoder.encode(Json.java:180)
	at net.starlark.java.lib.json.Json$Encoder.encode(Json.java:207)
	at net.starlark.java.lib.json.Json$Encoder.access$100(Json.java:103)
	at net.starlark.java.lib.json.Json.encode(Json.java:96)
	at jdk.internal.reflect.GeneratedMethodAccessor221.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at net.starlark.java.eval.MethodDescriptor.call(MethodDescriptor.java:162)
	at net.starlark.java.eval.BuiltinFunction.fastcall(BuiltinFunction.java:77)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:619)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:672)
	at net.starlark.java.eval.Eval.eval(Eval.java:489)
	at net.starlark.java.eval.Eval.execAssignment(Eval.java:109)
	at net.starlark.java.eval.Eval.exec(Eval.java:268)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
	at net.starlark.java.eval.StarlarkFunction.fastcall(StarlarkFunction.java:191)
	at net.starlark.java.eval.Starlark.fastcall(Starlark.java:619)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:672)
	at net.starlark.java.eval.Eval.eval(Eval.java:489)
	at net.starlark.java.eval.Eval.evalCall(Eval.java:630)
	at net.starlark.java.eval.Eval.eval(Eval.java:489)
	at net.starlark.java.eval.Eval.execReturn(Eval.java:249)
	at net.starlark.java.eval.Eval.exec(Eval.java:288)
	at net.starlark.java.eval.Eval.execStatements(Eval.java:82)
	at net.starlark.java.eval.Eval.execFunctionBody(Eval.java:66)
```

If I trim down the build graph this doesn't happen.

## Reproduction steps

```
git clone https://github.com/envoyproxy/envoy-mobile.git
cd envoy-mobile
git submodule update --init --recursive
git checkout rules_xcodeproj
bazel run //examples/swift/async_await:xcodeproj
```

Here's what the rules_xcodeproj integration looks like: https://github.com/envoyproxy/envoy-mobile/compare/rules_xcodeproj

If I trim down the dependency graph it works: https://github.com/envoyproxy/envoy-mobile/compare/rules_xcodeproj...rules_xcodeproj_minimal

In both those branches, bazel build //examples/swift/async_await:app works

## Expected behavior

Running the xcodeproj rule shouldn't use excessive memory.

## Versions
- Bazel: tested with 5.0.0 & 5.1.0
- rules_xcodeproj: 0.2.0
- rules_apple: 0.33.0
- rules_swift: https://github.com/bazelbuild/rules_swift/commits/a81f40700f1ba45034465a82673f46bd2631be62
- Xcode: 13.2.1
- Processor: M1 Max
<!-- Any others you feel are important to note -->

## Additional context
<!-- Add any other context about the problem here. -->

--- a/test/fixtures/cc/spec.json
+++ b/test/fixtures/cc/spec.json
@@ -259,8 +259,8 @@
             "configuration": "darwin_x86_64-fastbuild-ST-d53d69b6b8c1",
             "dependencies": [
                 "//examples/cc/lib:lib_impl darwin_x86_64-fastbuild-ST-d53d69b6b8c1",
-                "//examples/cc/lib2:lib_impl darwin_x86_64-fastbuild-ST-d53d69b6b8c1",
-                "@examples_cc_external//:lib_impl darwin_x86_64-fastbuild-ST-d53d69b6b8c1"
+                "@examples_cc_external//:lib_impl darwin_x86_64-fastbuild-ST-d53d69b6b8c1",
+                "//examples/cc/lib2:lib_impl darwin_x86_64-fastbuild-ST-d53d69b6b8c1"
             ],
             "frameworks": [],
             "info_plist": null,

--- a/xcodeproj/internal/target.bzl
+++ b/xcodeproj/internal/target.bzl
@@ -270,7 +270,8 @@ def _xcode_target(
         links: A `list` of file paths for libraries that the target links
             against.
         info_plist: A value as returned by `files.file_path()` or `None`.
-        dependencies: A `list` of `id`s of targets that this target depends on.
+        dependencies: A `depset` of `id`s of targets that this target depends
+            on.
         outputs: The value returned from `targets.get_outputs()`.
 
     Returns:
@@ -303,7 +304,7 @@ def _xcode_target(
         ),
         links = [file_path_to_dto(fp) for fp in links],
         info_plist = file_path_to_dto(info_plist),
-        dependencies = dependencies,
+        dependencies = dependencies.to_list(),
         outputs = outputs,
     ))
 
@@ -1174,17 +1175,23 @@ def _skip_target(*, target, transitive_infos):
     )
 
 def _process_dependencies(*, attrs_info, transitive_infos):
-    return [
-        dependency
-        for dependency in flatten([
+    direct_dependencies = []
+    transitive_dependencies = []
+    for attr, info in transitive_infos:
+        if not (not attrs_info or
+                info.target_type in attrs_info.xcode_targets.get(attr, [None])):
+            continue
+        if info.target:
+            direct_dependencies.append(info.target.id)
+        else:
             # We pass on the next level of dependencies if the previous target
             # didn't create an Xcode target.
-            [info.target.id] if info.target else info.dependencies
-            for attr, info in transitive_infos
-            if (not attrs_info or
-                info.target_type in attrs_info.xcode_targets.get(attr, [None]))
-        ])
-    ]
+            transitive_dependencies.append(info.dependencies)
+
+    return depset(
+        direct_dependencies,
+        transitive = transitive_dependencies,
+    )
 
 def _process_defines(*, cc_info, build_settings):
     if cc_info and build_settings != None:

